File: processor-service/internal/services/transfer_service.go
package services

import (
	"context"
	"fmt"
	"log/slog"
	"strings"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"

	"github.com/i0dk1/NeoWallet/processor-service/internal/apperror"
	"github.com/i0dk1/NeoWallet/processor-service/internal/clients"
	"github.com/i0dk1/NeoWallet/processor-service/internal/models"
	"github.com/i0dk1/NeoWallet/processor-service/internal/repository"
)

type TransferService struct {
	repo          *repository.Repository
	accountsClient *clients.AccountsClient
}

func New(repo *repository.Repository, accountsClient *clients.AccountsClient) *TransferService {
	return &TransferService{
		repo:           repo,
		accountsClient: accountsClient,
	}
}

func (s *TransferService) Transfer(ctx context.Context, req models.TransferRequest) (*models.TransferResponse, error) {
	slog.Info("iniciando transferencia",
		"sender_id", req.SenderID,
		"receiver_id", req.ReceiverID,
		"amount", req.Amount.String(),
	)

	if req.SenderID == req.ReceiverID {
		return nil, apperror.ErrSelfTransferNotAllowed
	}

	if err := validarMonto(req.Amount); err != nil {
		return nil, err
	}

	// validar que sender existe y tiene fondos
	sender, err := s.accountsClient.GetUser(ctx, req.SenderID)
	if err != nil {
		slog.Error("error verificando remitente", "error", err)
		return nil, apperror.ErrInternal
	}
	if sender == nil {
		return nil, apperror.ErrUserNotFound
	}
	if sender.Balance.LessThan(req.Amount) {
		return nil, apperror.ErrInsufficientFunds
	}

	// validar que receiver existe
	receiver, err := s.accountsClient.GetUser(ctx, req.ReceiverID)
	if err != nil {
		slog.Error("error verificando destinatario", "error", err)
		return nil, apperror.ErrInternal
	}
	if receiver == nil {
		return nil, apperror.ErrUserNotFound
	}

	// generar transaction_id
	txID := uuid.New()

	// crear transacción en estado PENDING
	_, err = s.repo.CreateTransaction(ctx, txID, req.SenderID, req.ReceiverID, req.Amount)
	if err != nil {
		slog.Error("error creando transacción en DB", "error", err)
		return nil, apperror.ErrInternal
	}

	slog.Info("transacción creada", "transaction_id", txID.String(), "status", "PENDING")

	// PASO 1: Débito al remitente
	debitResp, err := s.accountsClient.UpdateBalance(ctx, models.UpdateBalanceRequest{
		UserID:         req.SenderID,
		Amount:         req.Amount,
		Operation:      "debit",
		IdempotencyKey: fmt.Sprintf("%s:debit", txID.String()),
	})
	if err != nil {
		if strings.Contains(err.Error(), "insufficient_funds") {
			s.repo.UpdateStatus(ctx, txID, "FAILED", "fondos insuficientes al momento del débito")
			return nil, apperror.ErrInsufficientFunds
		}
		slog.Error("error en débito", "error", err, "transaction_id", txID.String())
		s.repo.UpdateStatus(ctx, txID, "FAILED", fmt.Sprintf("error en débito: %v", err))
		return nil, apperror.ErrInternal
	}

	s.repo.UpdateStatus(ctx, txID, "DEBITED", "")
	slog.Info("débito exitoso", "transaction_id", txID.String(), "previo", debitResp.PreviousBalance.String(), "nuevo", debitResp.NewBalance.String())

	// PASO 2: Crédito al destinatario
	_, err = s.accountsClient.UpdateBalance(ctx, models.UpdateBalanceRequest{
		UserID:         req.ReceiverID,
		Amount:         req.Amount,
		Operation:      "credit",
		IdempotencyKey: fmt.Sprintf("%s:credit", txID.String()),
	})
	if err != nil {
		slog.Error("error en crédito, iniciando compensación", "error", err, "transaction_id", txID.String())

		// COMPENSACIÓN: devolver el monto al remitente
		_, compErr := s.accountsClient.UpdateBalance(ctx, models.UpdateBalanceRequest{
			UserID:         req.SenderID,
			Amount:         req.Amount,
			Operation:      "credit",
			IdempotencyKey: fmt.Sprintf("%s:compensate", txID.String()),
		})
		if compErr != nil {
			slog.Error("ERROR CRÍTICO: falló la compensación", "error", compErr, "transaction_id", txID.String())
			s.repo.UpdateStatus(ctx, txID, "FAILED", fmt.Sprintf("falló compensación: %v", compErr))
			return nil, apperror.ErrTransferFailed
		}

		s.repo.UpdateStatus(ctx, txID, "ROLLED_BACK", fmt.Sprintf("crédito falló: %v", err))
		slog.Warn("transferencia compensada (ROLLED_BACK)", "transaction_id", txID.String())
		return nil, apperror.ErrTransferFailed
	}

	s.repo.UpdateStatus(ctx, txID, "COMPLETED", "")
	slog.Info("transferencia completada", "transaction_id", txID.String(), "status", "COMPLETED")

	return &models.TransferResponse{
		TransactionID: txID,
		Status:        "COMPLETED",
	}, nil
}

func (s *TransferService) GetTransactionHistory(ctx context.Context, userID int) ([]models.TransactionHistoryItem, error) {
	// verificar que el usuario existe
	user, err := s.accountsClient.GetUser(ctx, userID)
	if err != nil {
		return nil, apperror.ErrInternal
	}
	if user == nil {
		return nil, apperror.ErrUserNotFound
	}

	return s.repo.GetTransactionsByUserID(ctx, userID)
}

func (s *TransferService) ReconcileStuckTransactions(ctx context.Context, timeoutMinutes int) {
	txs, err := s.repo.GetStuckTransactions(ctx, timeoutMinutes)
	if err != nil {
		slog.Error("error en reconciliación", "error", err)
		return
	}

	for _, tx := range txs {
		slog.Warn("transacción atascada detectada", "transaction_id", tx.TransactionID.String(), "status", tx.Status)

		switch tx.Status {
		case "PENDING":
			// nunca se debitaron fondos; simplemente marcar como fallida
			s.repo.UpdateStatus(ctx, tx.TransactionID, "FAILED", "transacción atascada en PENDING")
		case "DEBITED":
			// el dinero fue debitado pero no acreditado. Intentar acreditar al receiver
			_, err := s.accountsClient.UpdateBalance(ctx, models.UpdateBalanceRequest{
				UserID:         tx.ReceiverID,
				Amount:         tx.Amount,
				Operation:      "credit",
				IdempotencyKey: fmt.Sprintf("%s:credit", tx.TransactionID.String()),
			})
			if err != nil {
				// compensar al sender
				s.accountsClient.UpdateBalance(ctx, models.UpdateBalanceRequest{
					UserID:         tx.SenderID,
					Amount:         tx.Amount,
					Operation:      "credit",
					IdempotencyKey: fmt.Sprintf("%s:compensate", tx.TransactionID.String()),
				})
				s.repo.UpdateStatus(ctx, tx.TransactionID, "ROLLED_BACK", "compensado en reconciliación")
			} else {
				s.repo.UpdateStatus(ctx, tx.TransactionID, "COMPLETED", "completado en reconciliación")
			}
		}
	}
}

func validarMonto(amount decimal.Decimal) error {
	if amount.LessThanOrEqual(decimal.Zero) {
		return apperror.NewValidationError("el monto debe ser positivo")
	}
	str := amount.String()
	parts := strings.Split(str, ".")
	if len(parts) == 2 && len(parts[1]) > 2 {
		return apperror.NewValidationError("el monto debe tener como máximo dos decimales")
	}
	return nil
}

File: processor-service/src/services/transferService.js
/**
 * Transfer Service - Patrón Saga
 * Lógica de negocio para transferencias P2P con manejo de compensación
 */

const { TransactionModel, TRANSACTION_STATUS } = require('../models/transactionModel');
const HttpClient = require('../utils/httpClient');

class TransferService {
    /**
     * Ejecuta una transferencia P2P entre usuarios
     * Implementa el patrón Saga para consistencia distribuida
     * 
     * Flujo:
     * 1. Validaciones iniciales
     * 2. Crear registro de transacción (PENDING)
     * 3. Debitar del sender (DEBITED)
     * 4. Acreditar al receiver (COMPLETED)
     * 5. Si falla paso 4, compensar (ROLLED_BACK)
     * 
     * @param {number} senderId - ID del usuario que envía
     * @param {number} receiverId - ID del usuario que recibe
     * @param {number} amount - Monto a transferir
     * @returns {Promise<Object>} Resultado de la transferencia
     */
    static async transfer(senderId, receiverId, amount) {
        let transaction = null;
        let debitSuccess = false;

        try {
            // ============================================
            // VALIDACIONES INICIALES
            // ============================================

            console.log('🔄 Iniciando transferencia P2P');
            console.log(`   Sender: ${senderId} → Receiver: ${receiverId} | Monto: $${amount}`);

            // Validar que no sea auto-transferencia
            if (senderId === receiverId) {
                const error = new Error('No puedes transferir dinero a ti mismo');
                error.code = 'SELF_TRANSFER_NOT_ALLOWED';
                error.statusCode = 400;
                throw error;
            }

            // Validar monto positivo
            if (!amount || amount <= 0) {
                const error = new Error('El monto debe ser positivo y mayor a cero');
                error.code = 'INVALID_AMOUNT';
                error.statusCode = 400;
                throw error;
            }

            // Validar precisión decimal (máximo 2 decimales)
            const decimalPlaces = (amount.toString().split('.')[1] || '').length;
            if (decimalPlaces > 2) {
                const error = new Error('El monto debe tener máximo 2 decimales');
                error.code = 'INVALID_DECIMAL_PRECISION';
                error.statusCode = 400;
                throw error;
            }

            // ============================================
            // VALIDAR EXISTENCIA DE USUARIOS Y FONDOS
            // ============================================

            console.log('🔍 Verificando sender...');
            const sender = await HttpClient.getAccount(senderId);

            if (sender.balance < amount) {
                const error = new Error('Fondos insuficientes');
                error.code = 'INSUFFICIENT_FUNDS';
                error.statusCode = 400;
                throw error;
            }

            console.log('🔍 Verificando receiver...');
            const receiver = await HttpClient.getAccount(receiverId);

            // ============================================
            // PASO 1: CREAR REGISTRO DE TRANSACCIÓN (PENDING)
            // ============================================

            console.log('📝 Creando registro de transacción (PENDING)...');
            transaction = await TransactionModel.create(senderId, receiverId, amount);
            console.log(`✅ Transacción creada: ID ${transaction.id}`);

            // ============================================
            // PASO 2: DEBITAR DEL SENDER (DEBITED)
            // ============================================

            console.log('💸 Debitando del sender...');
            try {
                await HttpClient.debitAccount(senderId, amount);
                debitSuccess = true;

                // Actualizar estado a DEBITED
                await TransactionModel.updateStatus(transaction.id, TRANSACTION_STATUS.DEBITED);
                console.log('✅ Débito exitoso');
            } catch (error) {
                console.error('❌ Error al debitar del sender:', error.message);

                // Marcar transacción como FAILED
                await TransactionModel.updateStatus(
                    transaction.id,
                    TRANSACTION_STATUS.FAILED,
                    `Fallo en débito: ${error.message}`
                );

                throw error;
            }

            // ============================================
            // PASO 3: ACREDITAR AL RECEIVER (COMPLETED)
            // ============================================

            console.log('💰 Acreditando al receiver...');
            try {
                await HttpClient.creditAccount(receiverId, amount);

                // Actualizar estado a COMPLETED
                await TransactionModel.updateStatus(transaction.id, TRANSACTION_STATUS.COMPLETED);
                console.log('✅ Crédito exitoso');
                console.log('🎉 Transferencia completada exitosamente');

            } catch (error) {
                console.error('❌ Error al acreditar al receiver:', error.message);

                // ============================================
                // COMPENSACIÓN: REVERTIR EL DÉBITO (ROLLBACK)
                // ============================================

                console.log('⚠️ Iniciando compensación...');
                try {
                    await HttpClient.creditAccount(senderId, amount);
                    console.log('✅ Compensación exitosa: débito revertido');

                    // Marcar como ROLLED_BACK
                    await TransactionModel.updateStatus(
                        transaction.id,
                        TRANSACTION_STATUS.ROLLED_BACK,
                        `Fallo en crédito: ${error.message}. Débito revertido.`
                    );
                } catch (compensationError) {
                    console.error('🚨 CRÍTICO: Fallo en compensación:', compensationError.message);

                    // Marcar como FAILED con información crítica
                    await TransactionModel.updateStatus(
                        transaction.id,
                        TRANSACTION_STATUS.FAILED,
                        `CRÍTICO: Fallo en crédito y compensación. Requiere revisión manual. Error: ${error.message}`
                    );

                    // Este es un caso crítico que requiere intervención manual
                    const criticalError = new Error('Error crítico: fallo en compensación. Requiere revisión manual.');
                    criticalError.code = 'CRITICAL_COMPENSATION_FAILED';
                    criticalError.statusCode = 500;
                    criticalError.transactionId = transaction.id;
                    throw criticalError;
                }

                // Re-lanzar el error original
                const rollbackError = new Error('Transferencia fallida y revertida');
                rollbackError.code = 'TRANSFER_ROLLED_BACK';
                rollbackError.statusCode = 500;
                rollbackError.originalError = error.message;
                throw rollbackError;
            }

            // ============================================
            // RETORNAR RESULTADO EXITOSO
            // ============================================

            return {
                success: true,
                message: 'Transferencia completada exitosamente',
                data: {
                    transaction_id: transaction.id,
                    sender_id: senderId,
                    receiver_id: receiverId,
                    amount: amount,
                    status: TRANSACTION_STATUS.COMPLETED,
                    timestamp: transaction.created_at
                }
            };

        } catch (error) {
            console.error('❌ Error en transferencia:', error.message);

            // Si el error ya tiene código, re-lanzarlo
            if (error.statusCode) {
                throw error;
            }

            // Error genérico
            const genericError = new Error('Error al procesar la transferencia');
            genericError.code = 'TRANSFER_PROCESSING_ERROR';
            genericError.statusCode = 500;
            genericError.originalError = error.message;
            throw genericError;
        }
    }

    /**
     * Obtiene el historial de transacciones de un usuario
     * @param {number} userId - ID del usuario
     * @returns {Promise<Object>} Historial formateado
     */
    static async getTransactionHistory(userId) {
        try {
            console.log(`📜 Consultando historial de usuario ${userId}`);

            // Verificar que el usuario existe
            await HttpClient.getAccount(userId);

            // Obtener transacciones
            const transactions = await TransactionModel.findByUserId(userId);

            // Formatear transacciones
            const formattedTransactions = transactions.map(tx => {
                const isSender = tx.sender_id === userId;
                const type = isSender ? 'sent' : 'received';
                const counterpartyId = isSender ? tx.receiver_id : tx.sender_id;

                return {
                    transaction_id: tx.id,
                    type: type,
                    amount: parseFloat(tx.amount),
                    counterparty_id: counterpartyId,
                    status: tx.status,
                    created_at: tx.created_at,
                    error_message: tx.error_message
                };
            });

            return {
                success: true,
                data: {
                    user_id: userId,
                    transaction_count: formattedTransactions.length,
                    transactions: formattedTransactions
                }
            };
        } catch (error) {
            console.error('❌ Error en getTransactionHistory:', error);
            throw error;
        }
    }

    /**
     * Obtiene estadísticas de transacciones (para debugging)
     * @returns {Promise<Object>} Estadísticas
     */
    static async getStatistics() {
        try {
            const counts = await TransactionModel.countByStatus();
            const recentTransactions = await TransactionModel.findAll(10);

            return {
                success: true,
                data: {
                    status_counts: counts,
                    recent_transactions: recentTransactions.map(tx => ({
                        id: tx.id,
                        sender_id: tx.sender_id,
                        receiver_id: tx.receiver_id,
                        amount: parseFloat(tx.amount),
                        status: tx.status,
                        created_at: tx.created_at
                    }))
                }
            };
        } catch (error) {
            console.error('❌ Error en getStatistics:', error);
            throw error;
        }
    }

    /**
     * Recarga saldo de un usuario
     * @param {number} userId - ID del usuario
     * @param {number} amount - Monto a recargar
     * @returns {Promise<Object>} Resultado de la recarga
     */
    static async recharge(userId, amount) {
        try {
            console.log(`💳 Recargando saldo para usuario ${userId}: $${amount}`);

            // Validar monto positivo
            if (!amount || amount <= 0) {
                const error = new Error('El monto debe ser positivo y mayor a cero');
                error.code = 'INVALID_AMOUNT';
                error.statusCode = 400;
                throw error;
            }

            // Validar precisión decimal (máximo 2 decimales)
            const decimalPlaces = (amount.toString().split('.')[1] || '').length;
            if (decimalPlaces > 2) {
                const error = new Error('El monto debe tener máximo 2 decimales');
                error.code = 'INVALID_DECIMAL_PRECISION';
                error.statusCode = 400;
                throw error;
            }

            // Verificar que el usuario existe
            const user = await HttpClient.getAccount(userId);

            // Acreditar el monto
            await HttpClient.creditAccount(userId, amount);

            // Obtener el nuevo saldo
            const updatedUser = await HttpClient.getAccount(userId);

            console.log(`✅ Recarga exitosa. Nuevo saldo: $${updatedUser.balance}`);

            return {
                success: true,
                message: 'Recarga completada exitosamente',
                data: {
                    user_id: userId,
                    amount: amount,
                    new_balance: parseFloat(updatedUser.balance),
                    previous_balance: parseFloat(user.balance)
                }
            };
        } catch (error) {
            console.error('❌ Error en recarga:', error.message);

            // Si el error ya tiene código, re-lanzarlo
            if (error.statusCode) {
                throw error;
            }

            // Error genérico
            const genericError = new Error('Error al procesar la recarga');
            genericError.code = 'RECHARGE_PROCESSING_ERROR';
            genericError.statusCode = 500;
            genericError.originalError = error.message;
            throw genericError;
        }
    }
}

module.exports = TransferService;
